### DISCUSSION TITLE
Kali Linux: Tech Key AC1200 USB Wifi Adapter; Realtek RTL88X2BU Driver Installation Successful

### DISCUSSION THREAD

**Discussion by Iterativehail27** [aircrack-ng/aircrack-ng#2381] [General]
Hello everyone, I was successful in installing the RTL88x2BU driver for this adapter:

[Techkey ac1200 wifi USB 3.0 adapter](https://www.amazon.com/dp/B07J65G9DD?psc=1&ref=ppx_yo2ov_dt_b_product_details)

On Kali linux 2022.4-live-amd64; as a live usb. 
<br>
<h2>Additional Information about my Install of Kali Linux:</h2>

```diff
> `Kali*kali: uname -a`

-> Linux kali 6.0.0-kali3-amd64 1 SMP PREEMPT_DYNAMIC Debian 6.0.7-1kali1 (2022-11-07) x86_64 GNU/Linux
```
<br>

<h2>To Install RTL88x2BU Driver Use These Commands as Follows:</h2>

```diff 
> `Kali*kali: sudo apt update && sudo apt upgrade`
> `Kali*kali: sudo apt install -y linux-headers-$(uname -r) build-essential bc dkms git libelf-dev`
> `Kali*kali: mkdir -p ~/src`
> `Kali*kali: cd ~/src`
> `Kali*kali: git clone https://github.com/morrownr/88x2bu-20210702.git`
> `Kali*kali: ls`
> `Kali*kali: cd /88x2bu-20210702`
> `Kali*kali: sudo ./install-driver.sh`
```
<h4>After installing I confirmed it supports <i><b>monitor</b></i> mode and <i><b> injection.</b></i></h4>
<br>

<h1>END NOTES:</h1>

I hope you all found this useful, if this helps you in some way please share your experience here or somewhere else if you have the time to follow-up with your successful installation of the RTL88x2BU driver on whatever NIC you may use.
<br>
Huge thank you to https://github.com/morrownr for providing the repository for the RTL88x2BU driver used in this install guide. 
Another thank you to Rokland Super Tech https://www.youtube.com/watch?v=Lx0j6KArKpA for his RTL8812AU install guide.
One more thank you to RTP Privacy Tech Tips https://www.youtube.com/watch?v=nehj6MJ4-hg&t=369s for her checklist/troubleshooting guide.

**Comment by ghost** [comment]
unsubscribe

**Comment by ZerBea** [comment]
I don't have such a device to test, so may I ask this question:
Does this driver support ioctl() system calls or does it depend on NETLINK only?

**Comment by kimocoder** [comment]
It's the same skeleton driver as rtl8812au uses.

**Comment by ZerBea** [comment]
@kimocoder , thanks for that information, Christian.
Maybe somebody can test wether this changes:
https://github.com/ZerBea/hcxdumptool/commit/1d9b463d5ee6da6bc9673c965ab6ccdd6bae3b3c
work on that driver or not.

For me, it  looks like Realtek drivers working better on 
```
pwrq.u.freq.m = freq * 100000
pwrq.u.freq.e = 1;
```
than on the old setting
```
pwrq.u.freq.m = freq
pwrq.u.freq.e = 6;
```
and I want to figure out why.
Output of hcxdumptool is very much appreciated:
```
$ hcxdumptool -i INTERFACE --check_driver
$ hcxdumptool -i INTERFACE -C
```
Please do net set monitor mode by iw or airmon-ng before, because this tools run NETLINK to set monitor mode.

**Reply by Iterativehail27** [reply]
``` diff
+> ──(kali㉿kali)-[~/src/88x2bu-20210702]
+>└─$ sudo hcxdumptool -i wlan1 --check_driver
->initialization of hcxdumptool 6.2.6 (depending on the capabilities of the device, this may take some time)...
->starting driver test...
->driver tests passed...
->all required ioctl() system calls are supported by driver

->terminating...
                                                                             
+>┌──(kali㉿kali)-[~/src/88x2bu-20210702]
+>└─$ sudo hcxdumptool -i wlan1 -C            
->initialization of hcxdumptool 6.2.6 (depending on the capabilities of the device, this may take some time)...
->wlan1 available frequencies, channels and tx power reported by driver:
->driver doesn't allow frequency scan
->driver doesn't allow frequency scan
->driver doesn't allow frequency scan
->driver doesn't allow frequency scan 

#It just prints the same output n, for any number of frequencies checked  

```
<h3> Hopefully that helps (notes below)</h3>
 I ended up having to install kali linux full (11.6gb) for live usb because the kali lite for live usb version did not have the tool hcxdumptool or the dependencies required and it utterly refused to install them (particularly pkg-config) even after running the recommended command. 
<br><br>

`apt-get install pkg-config libcurl4-openssl-dev libssl-dev zlib1g-dev make gcc`

 As a student of linux say if I want to learn how to install dependencies & understand linux as a whole better, do you think 
https://tldp.org/HOWTO/Software-Building-HOWTO.html and https://www.linuxfromscratch.org/lfs/view/stable/ 
are good resources or would you recommend something else or in addition to? I'd like to come back to this some day and be able to correctly install hcxdumptool for the kali lite version.

**Comment by ZerBea** [comment]
Great, thanks for that test and the information.
I'm interested in this Realtek driver test, to decide if it's worth it to add NETLINK support to hcxdumptool (direct via NETLINK API and not via libnl). I'm not a friend of NETLINK, because I prefer ioctl() system calls.
E.g. a simple ioctl(fd_usb, USBDEVFS_RESET, 0) will do an entire reset of an USB device.

hcxdumptool 6.2.6 is an older version (and the scan engine was still experimental).
I added the latest Realtek modifications 5 days ago:
https://github.com/ZerBea/hcxdumptool/commit/1d9b463d5ee6da6bc9673c965ab6ccdd6bae3b3c
for a better support of this driver (which is the best one of Realtek, I've ever seen - again thanks to Christian who provided the driver):
https://github.com/kimocoder/realtek_rtwifi/issues/34
It opened the door to cheap Realtek devices.

Studying Linux is very good and you can learn a lot. I'm on Linux since Buzz (1996) and I'm still learning (every day).
My favorite distribution is Arch Linux, running on the main systems (desktop and notebook)
and Debian sid,  because Arch dropped armv6 on Raspberry Pi first generation.
A Raspberry is really a tiny system:
https://github.com/ZerBea/hcxdumptool/wiki/Penetration-testing-system-1
 
For me (coder, signals and crypto analyst) the best way to learn Linux was/is/will be the documentation of the Linux Kernel on
https://docs.kernel.org/
and of course the Arch wiki
https://wiki.archlinux.org/
as well as the Debian wiki
https://wiki.debian.org

I don't like pre-configured systems installed by GUI installer. Doing this step by step by hand you'll learn a lot. And as benefit, you have a system that is exactly doing what you want to do (customized to meet your individual needs).

If you want to learn 802.11, too, I recommend the aircrack-ng suite. It is well documented and easy to use.
An absolute must have for carrying out analyses is tshark and Wireshark.
Basic information about 802.11 frames e.g. is here:
https://mrncciew.com
all frames are explained in detail, e.g.:
https://mrncciew.com/2014/10/08/802-11-mgmt-beacon-frame/
https://mrncciew.com/2014/10/28/802-11-mgmt-association-reqresponse/

A must have is 802.11 Wireless Networks: The Definitive Guide:
https://www.oreilly.com/library/view/80211-wireless-networks/0596100523/

**Reply by Iterativehail27** [reply]
Oh you're very welcome, happy to help. Hey and thank you for all of these good pointers with links and examples, I honestly couldn't ask for better help getting me started in the right direction to build from the floor up, I'm currently studying for the CCNA so that 802.11 wireless networks resource should fit right in to augment my understanding of wireless networks as well as all of the other linked resources for linux administration. You've been a big help, thanks again, let me know in the future if you need any additional help with something.

**Comment by ZerBea** [comment]
Thanks every help is welcome and you're welcome, too.
If it is allowed here (this is aircrack-ng discussions) to share Linux and 802.11 basic information, feel to ask.

**Reply by Mister-X-** [reply]
It is allowed, that's what this forum is meant for :)

**Comment by ZerBea** [comment]
Great, thanks.

**Comment by kimocoder** [comment]
Yeah, Im currently travelling but I will email you when I get back home, Mike 🥇

**Comment by ZerBea** [comment]
Thanks. Have a nice trip.
BTW:
I'm analyzing NETLINK messages from user space to driver and back.
Wireshark is a great tool to monitor what's going on there (e.g. debugging own messages):
```
$ sudo ip link add nlmon0 type nlmon
$ sudo ip link set dev nlmon0 up
``` 
First impression: It is much better to use NETLINK API ((faster, less overhead, full control)) than libnl.
We only need 2 sockets - that's all.

first socket to set interface  up and down: similar to ioctl(SIOCSIFFLAGS)
socke1(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE)

second socket to control the device: similar to ioctl(SIOCSIWMODE),  ioctl(SIOCSIWFREQ), ...
socket2(AF_NETLINK, SOCK_RAW | SOCK_CLOEXEC, NETLINK_GENERIC);

If Realtek decide to drop ioctl() support, we can use NETLINK API directly via this 2 sockets.
That means: no more "iw", no more "ip link" and no more libnl.
Luckily it doesn't look like they will do that (https://github.com/kimocoder/realtek_rtwifi).
I just ordered some cheap Realtek devices (< 9€) to dive deeper into the code of that driver. I think it is worth to do that.

**Comment by cregory777** [comment]
i cant seem to enable ipv6 and when i modprobe no such directory found i am sure is diff directory than rtl8812au can you please help
